src/ast/passes/semantic_analyser.cpp

```diff
@@ -1529,30 +1529,71 @@ Probe *SemanticAnalyser::get_probe_from_scope(Scope *scope,
 
 void SemanticAnalyser::visit(Map &map)
 {
-  MapKey key;
+  SizedType key = CreateNone();
+
+  if (map.key_expr) {
+    Visit(map.key_expr);
+    Expression *expr = map.key_expr;
+
+    if (expr->type.IsTupleTy()) {
+      // We can only do this expression and size type manipulation
+      // if we have access to the underlying expressions in the tuple ast
+      // e.g. @map[(1, 2)] vs $a = (1, 2) @map[$a]
+      auto tuple_ast = dynamic_cast<Tuple *>(expr);
+      if (tuple_ast) {
+        std::vector<SizedType> elements;
+        for (auto &elem : tuple_ast->elems) {
+          Visit(elem);
+
+          // If elem type is none that means that the tuple contains some
+          // invalid cast (e.g., (0, (aaa)0)). In this case, skip this.
+          // Cast already emits the error.
+          if (elem->type.IsNoneTy() || elem->type.GetSize() == 0)
+            return;
+
+          // Insert a cast to 64 bits if needed by injecting
+          // a cast into the ast.
+          if (elem->type.IsIntTy() && elem->type.GetSize() < 8) {
+            Expression *cast = ctx_.make_node<Cast>(elem->type.IsSigned()
+                                                        ? CreateInt64()
+                                                        : CreateUInt64(),
+                                                    elem,
+                                                    map.loc);
```

Comment:
This only works two-levels deep as well. This script breaks it:
```
BEGIN { @map[(1, ((uint16)2,3))] = 1; }
```

But see my comment above - is it even necessary?

Reply:
I actually think this isn't necessary - it was here before this PR but I yanked out this cast as we can just set the map key integer type to Int64 and everything works fine.